[main] feat(moe): Support gated delta net for Qwen3-Next (1/4)

yuzhongw-nvidia opened the pull request:
# What does this PR do ?
<!-- Add a one line overview of what this PR aims to accomplish. -->

[MR to dev](https://gitlab-master.nvidia.com/ADLR/megatron-lm/-/merge_requests/4070#note_41285641).

[Design doc](https://docs.google.com/document/d/1mGJDsqw4eej3965Ks65JQXuAqC8c_4Msnkm3btToHU4/edit?tab=t.0#heading=h.3ol9bs4x7jpw)

Qwen3-Next functionality PRs.
* https://github.com/NVIDIA/Megatron-LM/pull/2751
* https://github.com/NVIDIA/Megatron-LM/pull/2752
* https://github.com/NVIDIA/Megatron-LM/pull/2753

Changes in this PR:
1. Gated Delta Net (GDN); 
2. num_floating_point_operations


New supported arguments for Qwen3-Next

```yaml
--attention-output-gate: true  # Qwen3-Next applies output gate to standard attention layers
--no-weight-decay-cond-type: qwen3_next  # Qwen3-Next applies weight decay to qk layernorm as a special case
--moe-shared-expert-gate: true  # Qwen3-Next have a shared expert gate
# GDN args
--linear-attention-type: gated_delta_net
--linear-attention-freq: 4  # 1 1 1 0 1 1 1 0 ..., or we can use ([1,1,1,0]*12) to represent
--linear-conv-kernel-dim: 4
--linear-key-head-dim: 128
--linear-value-head-dim: 128
--linear-num-key-heads: 16
--linear-num-value-heads: 32
```

New for Qwen3-Next, but already supported in MCore yet

```yaml
--rotary-percent: 0.25  # Qwen3-Next applies RoPE only to the first 25% of position dimensions
--rotary-base: 10000000  # Qwen3-Next uses a larger rotary base
--apply-layernorm-1p: true  # Zero-Centered RMSNorm for Qwen3-Next
--max-position-embeddings: 262144  # Qwen3-Next-80B-A3B supports up to 256k sequence length
--moe-shared-expert-intermediate-size: 512  # Qwen3-Next have a shared expert
# MTP: Qwen3-Next has one MTP layer
--mtp-num-layers: 1
--mtp-loss-scaling-factor: 0.1
```

LM loss curve with the training dataset of Qwen3 are as below (GBS=256, seq_len=4096, TP1 is in green, TP2 is in blue).

[wandb url](https://wandb.ai/megatron-core-moe-dev/qwen3-next-benchmarking-baseline/workspace)

<img width="1438" height="813" alt="image" src="https://github.com/user-attachments/assets/9a00586e-51a8-4771-a50e-435fe0501c16" />
<img width="1443" height="818" alt="image" src="https://github.com/user-attachments/assets/2ce89c70-14b7-4185-9c71-d424921fa360" />
<img width="1428" height="818" alt="image" src="https://github.com/user-attachments/assets/b077e908-a2ba-4f87-8c94-e2dc40284c3f" />
<img width="1450" height="825" alt="image" src="https://github.com/user-attachments/assets/5e7193d4-0c3b-4444-937a-1898a2021d6b" />




:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 1): Add PR label `Expert Review`

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

jaredcasper commented on CODEOWNERS:
Why is this change here?

yuzhongw-nvidia commented:
> What tests have been done to validate the implementation of the gated delta net is correct?

Hi @jaredcasper , we have do several things to validate the implementation.
1. We have a [UT](https://github.com/yuzhongw-nvidia/Megatron-LM/blob/qwen3next/tests/unit_tests/ssm/test_gated_delta_net.py) to verify the implementation of GDN and parallel GDN.
2. We have used [the comparing tool in Megatron-Bridge](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/examples/conversion/README.md#comparepy---forward-pass-comparison) to verify the alignment with HF implementation.
3. We have run a E2E training test and see the loss curve makes sense. ([wandb](https://wandb.ai/megatron-core-moe-dev/qwen3-next-benchmarking-baseline/workspace))

In the next few weeks, we will further add a functional test to guarantee the long-term correctness of Qwen3-Next.

yuzhongw-nvidia commented:
> > In the next few weeks, we will further add a functional test to guarantee the long-term correctness of Qwen3-Next.
> 
> In the meantime do you think it makes sense to split this up and get the other features in?

Do you mean covering multiple features in a single functional test? It will be excellent as long as it is easy to maintain.

jaredcasper commented:
> Do you mean covering multiple features in a single functional test? It will be excellent, as long as we do not have too many features in one test case making it hard to maintain.

No, I mean there are 7 changes listed in the PR, do they all need to be in one PR? I'm saying while some of the features are getting tests added for them does it make sense to open PRs for some of the others that have been tested.

ericharper commented:
@yuzhongw-nvidia , can you fix conflicts and pull tot?

yuzhongw-nvidia commented:
> @yuzhongw-nvidia , can you fix conflicts and pull tot?

Yes, I have rebased main and fixed the conflicts.

jaredcasper commented on megatron/core/models/gpt/linear_attention_module_specs.py:
```diff
@@ -0,0 +1,39 @@
+) -> ModuleSpec:
+    """Helper function to get module spec for Linear Attention"""
+    rms_norm = normalization == "RMSNorm"
+    if linear_attention_type == "mamba":
```
Why can GPT models now have mamba attention? So what's the difference now between a GPT model and a Mamba Model? GPT Model is just more confusing?

yuzhongw-nvidia commented on megatron/core/models/gpt/gpt_layer_specs.py:
BTW, although we all think it is necessary to refactor gpt_layer_spec, but I'm not sure whether it should be in this PR. We could also fallback to the previous version and make the least changes to support Qwen3-Next's layer spec, and leave the refactor as our future work in another PR.

yuzhongw-nvidia replied:
> This spec functionality was never meant to handle the abstraction and flexibility needed for modern model structures. To try to use it for such purposes is forcing it to do something it wasn't designed to do and is just confusing.

Yes, I think I understand your point now. Let's have a new module to handle these things in the next PR.

yuzhongw-nvidia replied:
Changes done. Please take a look at the current version. @jaredcasper

BestJuly commented on megatron/training/training.py:
```diff
@@ -683,8 +772,15 @@ def pretrain(
+        model_provider,
+        model_type,
+        checkpointing_context=checkpointing_context,
+        no_wd_decay_cond=no_wd_decay_cond,
```
This should change the API of function `setup_model_and_optimizer` which is not included in this PR.

AllenShow commented:
Could you please implement supporting context parallel for qwen3-next?

rj42 commented:
Hi, great work, thank you for this code!

I've been using this code for my training and everything worked fine with `tp=1`, but with `tp > 1` things broke. I also tried the conversion tool from megatron-bridge ([compare.py ](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/9def9feeb7e0e3746d25e0277bf0435eea604b7f/examples/conversion/compare_hf_and_megatron/compare.py#L23) + `torchrun --nproc_per_node=2 compare.py --hf_model_path ./my-tiny-qwen/ --prompt "а а а а а а а а" --enable_debug_hooks --tp 2`) and encountered the same issue there.

I dug into the code and it seems like this part might not be correct for `tp > 1`:
[# Split, reorder, and reshape the tensor into q, k, v, gate, beta, alpha](https://github.com/NVIDIA/Megatron-LM/pull/1989/files#diff-82d2517679ca1af26f3a1ef6653a83ddf73c00c04ff3d1c009c2fa8728daa9b3R313) in `megatron/core/ssm/gated_delta_net.py`

The issue is that when `in_proj` weights are stored in packed format `[Q_all | K_all | V_all | Z_all | Beta_all | Alpha_all]`, the `ColumnParallelLinear` splits the output dimension evenly across TP ranks. This means each rank gets a slice of the concatenated tensor, not a proper per-component split.

For example, with tp=2 and total dim=5152:
- TP Rank 0 gets rows 0-2575: contains `[Q_all | K_all | V_partial]`
- TP Rank 1 gets rows 2576-5151: contains `[V_partial | Z_all | Beta_all | Alpha_all]`

But the split logic assumes each rank already has `[QKV_local | Z_local | Beta_local | Alpha_local]`, which doesn't match the actual data layout.

With this not-so-pretty fix, the tests pass: [link](https://github.com/rj42/Megatron-LM/commit/6b1a9513e0651e57547521db25b26eab344adfae#diff-82d2517679ca1af26f3a1ef6653a83ddf73c00c04ff3d1c009c2fa8728daa9b3R335).

This also led me to rewrite and slightly simplify the structure:

1. Split `in_proj` into two separate projections: `qkvz_proj` and `ba_proj`
2. Split the single `conv1d` into three separate ones: `q_conv1d`, `k_conv1d`, `v_conv1d`
3. Keep the interleaved weight format (same as HF) instead of merging into packed format — this way ColumnParallelLinear naturally splits by head groups, and each TP rank gets complete heads with all components (Q, K, V, Z, Beta, Alpha)
[megatron](https://github.com/rj42/Megatron-LM/commit/2b35273528581a2b60862605e322e8b6349f5f63) + [mbridge](https://github.com/rj42/Megatron-Bridge/commit/05f3056958d895c1f72858d0234887eedacbb229#diff-268dffa95fe65c8846135cf010c5a8d782365a329975c436dfcec49a6842fcf8) links for this realisation.

yuzhongw-nvidia commented:
> Hi, great work, thank you for this code!
> 
> I've been using this code for my training and everything worked fine with `tp=1`, but with `tp > 1` things broke. I also tried the conversion tool from megatron-bridge ([compare.py ](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/9def9feeb7e0e3746d25e0277bf0435eea604b7f/examples/conversion/compare_hf_and_megatron/compare.py#L23) + `torchrun --nproc_per_node=2 compare.py --hf_model_path ./my-tiny-qwen/ --prompt "а а а а а а а а" --enable_debug_hooks --tp 2`) and encountered the same issue there.
> 
> I dug into the code and it seems like this part might not be correct for `tp > 1`: [# Split, reorder, and reshape the tensor into q, k, v, gate, beta, alpha](https://github.com/NVIDIA/Megatron-LM/pull/1989/files#diff-82d2517679ca1af26f3a1ef6653a83ddf73c00c04ff3d1c009c2fa8728daa9b3R313) in `megatron/core/ssm/gated_delta_net.py`
> 
> The issue is that when `in_proj` weights are stored in packed format `[Q_all | K_all | V_all | Z_all | Beta_all | Alpha_all]`, the `ColumnParallelLinear` splits the output dimension evenly across TP ranks. This means each rank gets a slice of the concatenated tensor, not a proper per-component split.
> 
> For example, with tp=2 and total dim=5152:
> 
> * TP Rank 0 gets rows 0-2575: contains `[Q_all | K_all | V_partial]`
> * TP Rank 1 gets rows 2576-5151: contains `[V_partial | Z_all | Beta_all | Alpha_all]`
> 
> But the split logic assumes each rank already has `[QKV_local | Z_local | Beta_local | Alpha_local]`, which doesn't match the actual data layout.
> 
> With this not-so-pretty fix, the tests pass: [link](https://github.com/rj42/Megatron-LM/commit/6b1a9513e0651e57547521db25b26eab344adfae#diff-82d2517679ca1af26f3a1ef6653a83ddf73c00c04ff3d1c009c2fa8728daa9b3R335).
> 
> This also led me to rewrite and slightly simplify the structure:
> 
> 1. Split `in_proj` into two separate projections: `qkvz_proj` and `ba_proj`
> 2. Split the single `conv1d` into three separate ones: `q_conv1d`, `k_conv1d`, `v_conv1d`
> 3. Keep the interleaved weight format (same as HF) instead of merging into packed format — this way ColumnParallelLinear naturally splits by head groups, and each TP rank gets complete heads with all components (Q, K, V, Z, Beta, Alpha)
>    [megatron](https://github.com/rj42/Megatron-LM/commit/2b35273528581a2b60862605e322e8b6349f5f63) + [mbridge](https://github.com/rj42/Megatron-Bridge/commit/05f3056958d895c1f72858d0234887eedacbb229#diff-268dffa95fe65c8846135cf010c5a8d782365a329975c436dfcec49a6842fcf8) links for this realisation.

Hi @rj42 , thanks for your report. We've treated this problem in `GatedDeltaNet.sharded_state_dict` ([code](https://github.com/yuzhongw-nvidia/Megatron-LM/blob/qwen3next/megatron/core/ssm/gated_delta_net.py#L427-L502)), so the issue you mentioned should not have happened. Not only that, we've checked the correctness for TP > 1 by

- Unit tests ([code](https://github.com/yuzhongw-nvidia/Megatron-LM/blob/qwen3next/tests/unit_tests/ssm/test_gated_delta_net.py#L149))
- E2E pretraining task (using megatron-bridge to convert the open-source checkpoint).

I believe there are some configuration problems both in your run and in the megatron-bridge comparison script. Could you please open an issue to further discuss about this problem? If you can share your configuration or launch command, it will be better. Thanks!

rj42 commented:
> > Hi, great work, thank you for this code!
> > I've been using this code for my training and everything worked fine with `tp=1`, but with `tp > 1` things broke. I also tried the conversion tool from megatron-bridge ([compare.py ](https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/9def9feeb7e0e3746d25e0277bf0435eea604b7f/examples/conversion/compare_hf_and_megatron/compare.py#L23) + `torchrun --nproc_per_node=2 compare.py --hf_model_path ./my-tiny-qwen/ --prompt "а а а а а а а а" --enable_debug_hooks --tp 2`) and encountered the same issue there.
> > I dug into the code and it seems like this part might not be correct for `tp > 1`: [# Split, reorder, and reshape the tensor into q, k, v, gate, beta, alpha](https://github.com/NVIDIA/Megatron-LM/pull/1989/files#diff-82d2517679ca1af26f3a1ef6653a83ddf73c00c04ff3d1c009c2fa8728daa9b3R313) in `megatron/core/ssm/gated_delta_net.py`
> > The issue is that when `in_proj` weights are stored in packed format `[Q_all | K_all | V_all | Z_all | Beta_all | Alpha_all]`, the `ColumnParallelLinear` splits the output dimension evenly across TP ranks. This means each rank gets a slice of the concatenated tensor, not a proper per-component split.
> > For example, with tp=2 and total dim=5152:
> > 
> > * TP Rank 0 gets rows 0-2575: contains `[Q_all | K_all | V_partial]`
> > * TP Rank 1 gets rows 2576-5151: contains `[V_partial | Z_all | Beta_all | Alpha_all]`
> > 
> > But the split logic assumes each rank already has `[QKV_local | Z_local | Beta_local | Alpha_local]`, which doesn't match the actual data layout.
> > With this not-so-pretty fix, the tests pass: [link](https://github.com/rj42/Megatron-LM/commit/6b1a9513e0651e57547521db25b26eab344adfae#diff-82d2517679ca1af26f3a1ef6653a83ddf73c00c04ff3d1c009c2fa8728daa9b3R335).
> > This also led me to rewrite and slightly simplify the structure:
> > 
> > 1. Split `in_proj` into two separate projections: `qkvz_proj` and `ba_proj`
> > 2. Split the single `conv1d` into three separate ones: `q_conv1d`, `k_conv1d`, `v_conv1d`
> > 3. Keep the interleaved weight format (same as HF) instead of merging into packed format — this way ColumnParallelLinear naturally splits by head groups, and each TP rank gets complete heads with all components (Q, K, V, Z, Beta, Alpha)
> >    [megatron](https://github.com/rj42/Megatron-LM/commit/2b35273528581a2b60862605e322e8b6349f5f63) + [mbridge](https://github.com/rj42/Megatron-Bridge/commit/05f3056958d895c1f72858d0234887eedacbb229#diff-268dffa95fe65c8846135cf010c5a8d782365a329975c436dfcec49a6842fcf8) links for this realisation.
> 
> Hi @rj42 , thanks for your report. We've treated this problem in `GatedDeltaNet.sharded_state_dict` ([code](https://github.com/yuzhongw-nvidia/Megatron-LM/blob/qwen3next/megatron/core/ssm/gated_delta_net.py#L427-L502)), so the issue you mentioned should not have happened. Not only that, we've checked the correctness for TP > 1 by
> 
> * Unit tests ([code](https://github.com/yuzhongw-nvidia/Megatron-LM/blob/qwen3next/tests/unit_tests/ssm/test_gated_delta_net.py#L149))
> * E2E pretraining task (using megatron-bridge to convert the open-source checkpoint).
> 
> I believe there are some configuration problems both in your run and in the megatron-bridge comparison script. Could you please open an issue to further discuss about this problem? If you can share your configuration or launch command, it will be better. Thanks!

Created an [issue](https://github.com/NVIDIA/Megatron-LM/issues/2624) and add test examples.

yuzhongw-nvidia commented:
/ok to test e5ba8dc2fa945a966330d9b97359e74677467461

yuzhongw-nvidia commented:
/ok to test 218f5621294b26dd593849a98f3484b837636df8

yeontaek commented on megatron/core/ssm/gated_delta_net.py:
```diff
@@ -0,0 +1,657 @@
+                    dtype=self.config.params_dtype,
+                    device=torch.cuda.current_device(),
+                ).uniform_(*self.A_init_range)
+                self.A_log.data.copy_(A)
```
Bug: A_log should store log(A), not A directly. 

Current code stores raw A values, but forward() calls self.A_log.exp() assuming A_log contains log-space values. 

Fix: self.A_log.data.copy_(torch.log(A))

Phlip79 commented:
/ok to test 2a31244

BestJuly commented on megatron/training/training.py:
```diff
@@ -388,29 +387,101 @@ def transformer_flops():
+                    f"current linear attention pattern: {args.linear_attention_freq}"
+                )
+            elif args.linear_attention_freq is None:
+                linear_attention_pattern = [1] * num_layers
```
```suggestion
                # This should be caught by config validation, but raise here as a safety check
                raise ValueError(
                    f"Linear attention type {args.experimental_attention_variant} is specified "
                    "but linear_attention_freq is None. "
                    "Please set linear_attention_freq to specify the LA/SDPA layer pattern."
                )
```

BestJuly commented on megatron/core/ssm/gated_delta_net.py:
```diff
@@ -0,0 +1,657 @@
+        """
+
+        if not HAVE_FLA:
+            raise ImportError("FLA is not installed. Please install it with `pip install fla`.")
```
```suggestion
            raise ImportError("FLA is not installed. Please install it with `pip install flash-linear-attention`.")
```

yuzhongw-nvidia commented on megatron/training/training.py:
```diff
@@ -392,29 +391,106 @@ def transformer_flops():
+            key_projection_size = args.kv_channels * args.num_query_groups
+            value_projection_size = args.kv_channels * args.num_query_groups
+            standard_self_attn_term = (
+                3
```
I'm sure it is correct because:
* I have checked carefully and make sure it is equivalent to the previous code.
* I have checked in an E2E test and find it outputs the same result as before.

I also add some comments to make it easy to understand.

I do not agree with the point "use expansion_factor to make this more readable". I know most of the people are confused by it because there is a `2x` in the expansion_factor:
```
        # - 2x: GEMMs of a particular size are stacked twice in the standard Transformer model
        #       architectures implemented in this codebase (e.g., h->ffn_h GEMM and ffn_h->h GEMM
        #       in MLP layer).
```
which is only immediately obvious for MLP. Could you tell me what does it mean in each term? Many colleagues and I finally understand it after a careful derivation, reflecting how truly unreadable it previously was. Furthermore, the previous version has a chaotic calculation sequence and lacked comments, rendering it completely unreadable. That's why I argue that we must rewrite it.

jaredcasper replied:
Different approaches, I thought the previous was much more readable. :) 

In any case, there is a large comment explaining expansion_factor, and the creation of the variable itself, but expansion_factor is not used any more (unless I'm missing its use?). Please clean this up if you are going to remove expansion_factor. The comments explaining expansion factor can be moved down here.

yuzhongw-nvidia commented:
/ok to test 3e1ca94b1e5a862c890d673b534c53b5b08f462a

yuzhongw-nvidia commented:
/ok to test 418be3ce003cfeb1d8b5da186de26430de836074

yuzhongw-nvidia commented:
/ok to test fd96ce8e2893bd6371c16795dba9aad009009121

yuzhongw-nvidia commented:
/ok to test 13d2cc9a944a3dba77fd84ab1cd066349b5d5f3c
